From 3d0905acd286a26dd59f00a4dfc78c5c95dfc019 Mon Sep 17 00:00:00 2001
From: Andrew Waterman <waterman@cs.berkeley.edu>
Date: Sun, 9 Oct 2016 14:59:10 -0700
Subject: [PATCH 09/10] Properly truncate sub-word values in __put_user

We need an explicit up-cast from char/short back to long, in order for
the sub-word quantity to be truncated.

h/t Richard W.M. Jones
---
 arch/riscv/include/asm/uaccess.h | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/riscv/include/asm/uaccess.h b/arch/riscv/include/asm/uaccess.h
index b7532de1..b139cbf 100644
--- a/arch/riscv/include/asm/uaccess.h
+++ b/arch/riscv/include/asm/uaccess.h
@@ -268,6 +268,7 @@ do {								\
 #define __put_user_asm(insn, x, ptr, err)			\
 do {								\
 	uintptr_t __tmp;					\
+	__typeof__(*(ptr)) __x = x;				\
 	__enable_user_access();					\
 	__asm__ __volatile__ (					\
 		"1:\n"						\
@@ -284,7 +285,7 @@ do {								\
 		"	" PTR " 1b, 3b\n"			\
 		"	.previous"				\
 		: "+r" (err), "=r" (__tmp), "=m" (*(ptr))	\
-		: "rJ" (x), "i" (-EFAULT));			\
+		: "rJ" (__x), "i" (-EFAULT));			\
 	__disable_user_access();				\
 } while (0)
 #else /* !CONFIG_MMU */
@@ -292,7 +293,7 @@ do {								\
 	__asm__ __volatile__ (					\
 		insn " %z1, %0"					\
 		: "=m" (*(ptr))					\
-		: "rJ" (x))
+		: "rJ" ((__typeof__(*(ptr))) x))
 #endif /* CONFIG_MMU */
 
 
-- 
2.10.2

